doc/internal/developer.md

```diff
@@ -0,0 +1,355 @@
+
+One of the things that's useful to do in a routine CI run is to
+generate a dummy snapshot release (as in, the release artefacts, like
+source and binary tarballs, documentation PDFs, etc.) to make sure this
+generation isn't broken.
+We don't currently do this, but we probably should.
+
+## Testing with unlocked dependencies
+
+Another useful thing to do is to see what happens if you build against
+the latest version of all the things you depend on.
+If that fails, it's probably something you want to know about, even if you
+don't do anything about it right away.
+
+There are at least four subcategories of this (submodules,
+dependencies on packages we control, dependencies on packages we
+don't, and external installed dependencies) that require different
+handling.
+Currently we don't do any of this, at least not as part of the CI runs.
+It isn't clear that GitHub Actions is the right venue for it, either.
+
+## Heterogeneity and monoculture
+
+One of the risks with containers and slickly packaged CI testing and
+whatnot is that your testing environment becomes a monoculture, and
+that makes your system brittle.
+Portability is less of a problem with Haskell code than in other
+contexts, but all the same we try to avoid this by testing on multiple
+OSes and multiple target CPUs.
+
+Currently we build on two supported Ubuntu LTS releases on x86, two
+MacOS releases, one x86 and one arm64, and Windows.
+It's possible we should add something else, maybe using our own runner.
+
+
+# Issues
+
+See [issue-labels.md](issue-labels.md) for documentation of the issue
+labels we use.
+
+## When to file an issue
+
+We do not require that every pull request correspond to an issue.
+It is fine if there are several pull requests before an issue is
+finally resolved (as long as what's happening is reasonably clear to
+passersby) and also fine to make pull requests without an issue.
+
+Create an issue if:
+
+- there's a problem that you're not going to get to right away;
+
+- there are complicated circumstances that should be written down,
+or there's a problem that someone might want to be able to find
+again easily later; or
+
+- there needs to be a discussion about what to do, in order to
+gather consensus or consult with stakeholders.
+
+Don't create an issue if you already have a patch, unless one of these
+circumstances applies or there's some other reason to specifically
+create an issue; just file a pull request.
+This is especially true for minor changes.
+Changes that are _not_ likely to be searched for again later should
+be kept out of the issue database to the extent reasonably possible.
+
+Note that if you are searching closed issues to try to find something
+you vaguely remember seeing go by, you may want to also search closed
+pull requests.
+Fortunately, on GitHub this is extremely easy. 
+
+
+# Documentation
+
+So far there is nothing that needs to be said here that isn't in
+[../../CONTRIBUTING.md](CONTRIBUTING).
+
```

Comment:
N.b. If we merge this before #2202 (likely), I'll update this section with a link to the `doc/README`.